File: swig/cpp/tests/changes.cpp
#include <iostream>
#include <memory>
#include <cassert>
#include <cstring>
#include <unistd.h>

#include "Session.hpp"

#define MAX_LEN 100

using namespace std;

const string module_name = "swig-test-cpp-changes";
const int LOW_BOUND = 10;
const int HIGH_BOUND = 20;

std::string get_test_name(int i)
{
    return "test-cpp-" + to_string(i);
}

std::string get_xpath(const std::string &test_name, const std::string &node_name)
{
    return "/" + module_name + ":cpp-changes/test-get[name='" + test_name + "']/" + node_name;
}

class NopCallback: public sysrepo::Callback {
public:
    int module_change(sysrepo::S_Session sess, const char *module_name, sr_notif_event_t event, void *private_ctx)
        {
            return SR_ERR_OK;
        }
};

void init_test(sysrepo::S_Session sess)
{
    sysrepo::S_Subscribe subs(new sysrepo::Subscribe(sess));
    sysrepo::S_Callback cb(new NopCallback());

    subs->module_change_subscribe(module_name.c_str(), cb, NULL, 0, SR_SUBSCR_DEFAULT | SR_SUBSCR_APPLY_ONLY);

    for (int i = LOW_BOUND; i < HIGH_BOUND; i++) {
        const auto xpath = get_xpath(get_test_name(i), "number");
        sysrepo::S_Val vset(new sysrepo::Val((int32_t)i, SR_INT32_T));
        sess->set_item(xpath.c_str(), vset);
    }

    sess->commit();
    sess->copy_config(module_name.c_str(), SR_DS_RUNNING, SR_DS_STARTUP);
    subs->unsubscribe();
}

void
clean_test(sysrepo::S_Session sess)
{
    string module_name_tmp(module_name);
    const string xpath = "/" + module_name_tmp + ":*";

    sess->delete_item(xpath.c_str());
    sess->commit();
    // sess.copy_config(module_name, sr_datastore_t.SR_DS_RUNNING, sr_datastore_t.SR_DS_STARTUP);
}

class DeleteCb: public sysrepo::Callback {
public:
    int module_change(sysrepo::S_Session sess, const char *module_name, sr_notif_event_t event, void *private_ctx)
        {
            char change_path[MAX_LEN];

            snprintf(change_path, MAX_LEN, "/%s:*", module_name);
            auto it = sess->get_changes_iter(&change_path[0]);
            auto change = sess->get_change_next(it);

            assert(SR_OP_DELETED == change->oper());
            assert(change->old_val()->data()->get_int32() == 10);
            return SR_ERR_OK;
        }
};

void
test_module_change_delete(sysrepo::S_Session sess)
{
    sysrepo::S_Subscribe subs(new sysrepo::Subscribe(sess));
    sysrepo::S_Callback cb(new DeleteCb());

    init_test(sess);
    subs->module_change_subscribe(module_name.c_str(), cb, NULL, 0, SR_SUBSCR_DEFAULT | SR_SUBSCR_APPLY_ONLY);

    const auto xpath = get_xpath(get_test_name(LOW_BOUND), "number");
    sess->delete_item(xpath.c_str());
    sess->commit();

    subs->unsubscribe();
}

class ModifyCb: public sysrepo::Callback {
public:
    int module_change(sysrepo::S_Session sess, const char *module_name, sr_notif_event_t event, void *private_ctx)
        {
            char change_path[MAX_LEN];
            snprintf(change_path, MAX_LEN, "/%s:*", module_name);
            auto it = sess->get_changes_iter(&change_path[0]);
            auto change = sess->get_change_next(it);

            assert(SR_OP_MODIFIED == change->oper());
            string xpath(change->new_val()->xpath());
            assert("/swig-test-cpp-changes:cpp-changes/test-get[name='test-cpp-10']/number" ==
                   xpath);
            assert(change->new_val()->data()->get_int32() == 42);
            return SR_ERR_OK;
        }
};

void
test_module_change_modify(sysrepo::S_Session sess)
{
    sysrepo::S_Subscribe subs(new sysrepo::Subscribe(sess));
    sysrepo::S_Callback cb(new ModifyCb());

    init_test(sess);

    subs->module_change_subscribe(module_name.c_str(), cb, NULL, 0, SR_SUBSCR_DEFAULT | SR_SUBSCR_APPLY_ONLY);

    const auto xpath = get_xpath(get_test_name(LOW_BOUND), "number");
    sysrepo::S_Val vset(new sysrepo::Val((int32_t)42, SR_INT32_T));
    sess->set_item(xpath.c_str(), vset);
    sess->commit();
    subs->unsubscribe();
}

class CreateCb: public sysrepo::Callback {
public:
    int module_change(sysrepo::S_Session sess, const char *module_name, sr_notif_event_t event, void *private_ctx)
        {
            char change_path[MAX_LEN];

            snprintf(change_path, MAX_LEN, "/%s:*", module_name);
            auto it = sess->get_changes_iter(&change_path[0]);
            auto change = sess->get_change_next(it);

            assert(SR_OP_CREATED == change->oper());
            string xpath(change->new_val()->xpath());
            assert("/swig-test-cpp-changes:cpp-changes/test-get[name='test-cpp-20']" ==
                   xpath);
            change = sess->get_change_next(it);
            assert(change->new_val()->data()->get_string() == string("test-cpp-20"));
            change = sess->get_change_next(it);
            assert(change->new_val()->data()->get_int32() == 42);

            return SR_ERR_OK;
        }
};

void
test_module_change_create(sysrepo::S_Session sess)
{
    sysrepo::S_Subscribe subs(new sysrepo::Subscribe(sess));
    sysrepo::S_Callback cb(new CreateCb());

    init_test(sess);

    subs->module_change_subscribe(module_name.c_str(), cb, NULL, 0, SR_SUBSCR_DEFAULT | SR_SUBSCR_APPLY_ONLY);

    const auto xpath = get_xpath(get_test_name(HIGH_BOUND), "number");
    sysrepo::S_Val vset(new sysrepo::Val((int32_t)42, SR_INT32_T));
    sess->set_item(xpath.c_str(), vset);
    sess->commit();

    subs->unsubscribe();
}

int
main(int argc, char **argv)
{
    sysrepo::S_Connection conn(new sysrepo::Connection("test changes"));
    sysrepo::S_Session sess(new sysrepo::Session(conn, SR_DS_RUNNING));

    clean_test(sess);
    test_module_change_delete(sess);
    clean_test(sess);
    test_module_change_modify(sess);
    test_module_change_create(sess);

    return 0;
}

File: swig/lua/tests/changes.lua
local sr = require("libsysrepoLua")
local lu = require('luaunit')

local MODULE_NAME = 'swig-test' -- name of module used for testing
local LOW_BOUND = 10                    -- for testing purposes, index of lowest 'eth' interface name
local HIGH_BOUND = 20                   -- for testing purposes, index of highest 'eth' interface name
local CHANGE_PATH = "/" .. MODULE_NAME .. ":*"
local xpath_if_fmt = "/swig-test:lua-changes/test[name='%s']/%s"

-- Helper functions
local function sleep()
  local _sleep_ignore = 0
  for i=1,100000000 do
    _sleep_ignore = i + _sleep_ignore
  end
end

function clean_slate(sess)

  local function _cb(session, module_name, event, private_ctx)
    print('clean cb')
    return tonumber(sr.SR_ERR_OK)
  end

  local xpath = "/" .. MODULE_NAME .. ":*"
  local subs = sr.Subscribe(sess)
  local wrap = sr.Callback_lua(_cb)
  subs:module_change_subscribe(MODULE_NAME, wrap, nil, 0, sr.SR_SUBSCR_APPLY_ONLY);

  print('cleaning up', xpath)

  sess:delete_item(xpath)
  sess:commit()
  sess:copy_config(MODULE_NAME, sr.SR_DS_RUNNING, sr.SR_DS_STARTUP)
  sleep()
end

local function get_test_name(i)
  return 'luatest'..'-'..tostring(i)
end

local function init_test(sess)
  local function _cb(session, module_name, event, private_ctx)
      print('init cb')
      return tonumber(sr.SR_ERR_OK)
  end

  local subs = sr.Subscribe(sess);
  local wrap = sr.Callback_lua(_cb)
  subs:module_change_subscribe(MODULE_NAME, wrap, nil, 0, sr.SR_SUBSCR_APPLY_ONLY);

  for i=LOW_BOUND, HIGH_BOUND do
    local val = sr.Val(i, sr.SR_INT32_T)
    local xpath = string.format(xpath_if_fmt, get_test_name(i), 'number')
    sess:set_item(xpath, val)
  end

  sess:commit()
  sess:copy_config(MODULE_NAME, sr.SR_DS_RUNNING, sr.SR_DS_STARTUP)
  sleep()
  subs:unsubscribe()
  sleep()
end

-- Suite start:
local conn = sr.Connection('main connection')
local sess = sr.Session(conn)

clean_slate(sess)

-- Test suite:
TestChanges = {}

function TestChanges:setUp()
  self.setup_test_num = 42
end

-- Clean-up after test has executed.
function TestChanges:tearDown()
  collectgarbage()
  collectgarbage()
end

function TestChanges:test_module_change()

  local set_num = 42
  local mod_num = 43
  local deletitionp = false;

  init_test(sess)

  local function _cb(session, module_name, event, private_ctx)
    print('test_module_change_cb', deletitionp)
    if not deletitionp then
      deletitionp = true
      return tonumber(sr.SR_ERR_OK)
    end

    local it = session:get_changes_iter(CHANGE_PATH);
    lu.assertNotIsNil(it)

    while true do
      local change = session:get_change_next(it)
      if (change == nil) then break end
      lu.assertEquals(change:oper(), sr.SR_OP_MODIFIED)
    end

    return tonumber(sr.SR_ERR_OK)
  end

  -- local conn, sess, subs = init_helper('change-modify')
  -- local wrap = sr.Callback_lua(_cb)
  -- subs:module_change_subscribe(MODULE_NAME, wrap, nil, 0, sr.SR_SUBSCR_APPLY_ONLY);

  local wrap = sr.Callback_lua(_cb)
  local subs = sr.Subscribe(sess)
  subs:module_change_subscribe(MODULE_NAME, wrap, nil, 0, sr.SR_SUBSCR_APPLY_ONLY);


  local val = sr.Val(set_num, sr.SR_INT32_T)
  local xpath = string.format(xpath_if_fmt, get_test_name(set_num), 'number')
  sess:set_item(xpath, val)
  sess:commit()
  sleep()

  local xpath = string.format(xpath_if_fmt, get_test_name(set_num), 'number')
  local val2 = sr.Val(mod_num, sr.SR_INT32_T)
  sess:set_item(xpath, val2)
  sess:commit()

  sleep()
  subs:unsubscribe()
  sleep()
end

function TestChanges:test_module_change_delete()

  local set_num = 44
  local deletitionp = false;

  init_test(sess)

  local function _cb(session, module_name, event, private_ctx)
    print('test_module_change_delete_cb', deletitionp)
    if not deletitionp then
      deletitionp = true
      return tonumber(sr.SR_ERR_OK)
    end

    local it = session:get_changes_iter(CHANGE_PATH);
    lu.assertNotIsNil(it)

    while true do
      local change = session:get_change_next(it)
      if (change == nil) then break end
      lu.assertEquals(change:oper(), sr.SR_OP_DELETED)
    end

    return tonumber(sr.SR_ERR_OK)
  end

  local wrap = sr.Callback_lua(_cb)
  local subs = sr.Subscribe(sess)
  subs:module_change_subscribe(MODULE_NAME, wrap, nil, 0, sr.SR_SUBSCR_APPLY_ONLY);

  local setval = sr.Val(set_num, sr.SR_INT32_T)
  local xpath = string.format(xpath_if_fmt, get_test_name(set_num), 'number')
  sess:set_item(xpath, setval)
  sess:commit()
  sleep()

  sess:delete_item(xpath)
  sess:commit()

  sleep()
  subs:unsubscribe()
  sleep()
end

local runner = lu.LuaUnit.new()
runner:setOutputType("tap")
local rc = runner:runSuite()
clean_slate(sess)
os.exit(rc)
